nixos/modules/services/databases/postgresql.nix

```diff
@@ -623,7 +623,46 @@ in
             TimeoutSec = 120;
 
             ExecStart = "${postgresql}/bin/postgres";
+
+            # Hardening
+            CapabilityBoundingSet = [ "" ];
+            DevicePolicy = "closed";
+            PrivateTmp = true;
+            ProtectHome = true;
+            ProtectSystem = "strict";
+            MemoryDenyWriteExecute = lib.mkDefault (cfg.settings.jit == "off");
+            NoNewPrivileges = true;
+            LockPersonality = true;
+            PrivateDevices = true;
+            PrivateMounts = true;
+            ProcSubset = "pid";
+            ProtectClock = true;
+            ProtectControlGroups = true;
+            ProtectHostname = true;
+            ProtectKernelLogs = true;
+            ProtectKernelModules = true;
+            ProtectKernelTunables = true;
+            ProtectProc = "invisible";
+            RemoveIPC = true;
+            RestrictAddressFamilies = [
+              "AF_INET"
+              "AF_INET6"
+              "AF_NETLINK" # used for network interface enumeration
+              "AF_UNIX"
+            ];
+            RestrictNamespaces = true;
+            RestrictRealtime = true;
+            RestrictSUIDSGID = true;
+            SystemCallArchitectures = "native";
+            SystemCallFilter = [
```

Comment:
I use plv8 extension and the database server crashes every time a query using plv8 function is called:

```
postgres[2325667]: [2325667] LOG:  server process (PID 2327131) was terminated by signal 31: Bad system call
```

Is there a way to determine which system calls are needed?

I tried

```nix
SystemCallFilter = [ "" ];
SystemCallLog = [ "@privileged ~@resources ~@system-service" ];
```

but it only made it crash with

```
Nov 10 15:43:06 azazel postgres[2336657]: #
Nov 10 15:43:06 azazel postgres[2336657]: # Fatal error in , line 0
Nov 10 15:43:06 azazel postgres[2336657]: # Check failed: 12 == (*__errno_location ()).
Nov 10 15:43:06 azazel postgres[2336657]: #
Nov 10 15:43:06 azazel postgres[2336657]: #
Nov 10 15:43:06 azazel postgres[2336657]: #
Nov 10 15:43:06 azazel postgres[2336657]: #FailureMessage Object: 0x7ffff76dee30
Nov 10 15:43:06 azazel postgres[2336657]: ==== C stack trace ===============================
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::base::debug::StackTrace::StackTrace()+0x16) [0x7f056f00e386]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(+0xb12e4a) [0x7f056ead7e4a]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(V8_Fatal(char const*, ...)+0x17a) [0x7f056e8aacea]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(+0x8de83d) [0x7f056e8a383d]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::MemoryAllocator::SetPermissionsOnExecutableMemoryChunk(v8::internal::VirtualMemory*, unsigned long, unsigned long, unsigned long)+0xfa) [0x7f056ea4222a]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::MemoryAllocator::AllocateAlignedMemory(unsigned long, unsigned long, unsigned long, v8::internal::AllocationSpace, v8::internal::Executability, void*, v8::internal::VirtualMemory*)+0x1f5) [0x7f056ea42505]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::MemoryAllocator::AllocateUninitializedChunkAt(v8::internal::BaseSpace*, unsigned long, v8::internal::Executability, unsigned long, v8::internal::PageSize)+0x96) [0x7f056ea42606]
Nov 10 15:43:06 azazel kernel: traps: postgres[387266] trap int3 ip:7f056e8a3b2e sp:7ffff76dee08 error:0 in plv8-3.2.3.so[7f056e7e7000+1b90000]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::MemoryAllocator::AllocatePage(v8::internal::MemoryAllocator::AllocationMode, v8::internal::Space*, v8::internal::Executability)+0x7e) [0x7f056ea42d8e]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::PagedSpaceBase::TryExpandImpl()+0x2c) [0x7f056ea56d0c]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::PagedSpaceBase::TryExpand(int, v8::internal::AllocationOrigin)+0x1f) [0x7f056ea58a0f]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::PagedSpaceBase::RawRefillLabMain(int, v8::internal::AllocationOrigin)+0x416) [0x7f056ea593d6]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::PagedSpaceBase::RefillLabMain(int, v8::internal::AllocationOrigin)+0x28) [0x7f056ea596c8]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment)+0x5e1) [0x7f056e9c6201]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::Factory::CodeBuilder::AllocateInstructionStream(bool)+0x7d) [0x7f056e9ac52d]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::Factory::CodeBuilder::BuildInternal(bool)+0x21b) [0x7f056e9ac95b]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::compiler::CodeGenerator::FinalizeCode()+0x16c) [0x7f056fe8ba7c]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::compiler::PipelineImpl::FinalizeCode(bool)+0x4ef) [0x7f056f1753ef]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::compiler::PipelineCompilationJob::FinalizeJobImpl(v8::internal::Isolate*)+0x70) [0x7f056f1763e0]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::OptimizedCompilationJob::FinalizeJob(v8::internal::Isolate*)+0x44) [0x7f056e8b8234]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::Compiler::FinalizeTurbofanCompilationJob(v8::internal::TurbofanCompilationJob*, v8::internal::Isolate*)+0x40b) [0x7f056e8be13b]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::OptimizingCompileDispatcher::InstallOptimizedFunctions()+0xc9) [0x7f056e8f24a9]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::StackGuard::HandleInterrupts()+0x1e0) [0x7f056e981590]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*)+0xb9) [0x7f056f3adef9]
Nov 10 15:43:06 azazel postgres[2336657]:     /nix/store/5nvfkms6x5nkaaq44w0xw7ph5dg051y6-postgresql-and-plugins-14.13/lib/plv8-3.2.3.so(+0xf1def6) [0x7f056eee2ef6]
Nov 10 15:43:06 azazel postgres[2336019]: [2336019] LOG:  server process (PID 2336657) was terminated by signal 5: Trace/breakpoint trap
Nov 10 15:43:06 azazel postgres[2336019]: [2336019] DETAIL:  Failed process was running: with recent_event as (select id from event where level = 'world' and "end" < current_date and "end" > '2022-01-01' order by "end" desc limit 2), past_team as (select *, 0.5 as coef, 6 as ceiling, position_in_class(age, 'ultraveteran') over (w order by score desc, time asc) as position_uv, count(nullif(is_in_class(age, 'ultraveteran'), false)) over w as limit_uv, position_in_class(age, 'superveteran') over (w order by score desc, time asc) as position_sv, count(nullif(is_in_class(age, 'superveteran'), false)) over w as limit_sv, position_in_class(age, 'veteran') over (w order by score desc, time asc) as position_v, count(nullif(is_in_class(age, 'veteran'), false)) over w as limit_v, position_in_class(age, 'open') over (w order by score desc, time asc) as position_o, count(nullif(is_in_class(age, 'open'), false)) over w as limit_o, position_in_class(age, 'junior') over (w order by score desc, time asc) as position_j, count(nullif(is_in_class(age, 'junior'), false)) over w as limit_j from team where event_id in
Nov 10 15:43:06 azazel postgres[2336019]: [2336019] LOG:  terminating any other active server processes
Nov 10 15:43:06 azazel node[2333506]: Error: Connection terminated unexpectedly
Nov 10 15:43:06 azazel node[2333506]:     at /nix/store/yq48bpc7clgjrybghnz5a57fnsw4scb4-wrcq-0-unstable-2024-10-22/node_modules/pg-pool/index.js:45:11
Nov 10 15:43:06 azazel node[2333506]:     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Nov 10 15:43:06 azazel node[2333506]:     at async file:///nix/store/yq48bpc7clgjrybghnz5a57fnsw4scb4-wrcq-0-unstable-2024-10-22/index.js:309:32
Nov 10 15:43:06 azazel postgres[2336665]: [2336665] FATAL:  the database system is in recovery mode
Nov 10 15:43:06 azazel node[2333506]: ::1 - - [10/Nov/2024:14:43:06 +0000] "GET /qualified HTTP/1.0" 500 -
Nov 10 15:43:06 azazel node[2333506]: error: the database system is in recovery mode
Nov 10 15:43:06 azazel node[2333506]:     at /nix/store/yq48bpc7clgjrybghnz5a57fnsw4scb4-wrcq-0-unstable-2024-10-22/node_modules/pg-pool/index.js:45:11
Nov 10 15:43:06 azazel node[2333506]:     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Nov 10 15:43:06 azazel node[2333506]:     at async PGStore._asyncQuery (/nix/store/yq48bpc7clgjrybghnz5a57fnsw4scb4-wrcq-0-unstable-2024-10-22/node_modules/connect-pg-simple/index.js:322:21)
Nov 10 15:43:06 azazel postgres[2336019]: [2336019] LOG:  all server processes terminated; reinitializing
Nov 10 15:43:06 azazel postgres[2336666]: [2336666] LOG:  database system was interrupted; last known up at 2024-11-10 14:42:55 GMT
Nov 10 15:43:06 azazel postgres[2336666]: [2336666] LOG:  database system was not properly shut down; automatic recovery in progress
Nov 10 15:43:06 azazel postgres[2336666]: [2336666] LOG:  redo starts at 2/9F5C7BE0
Nov 10 15:43:06 azazel postgres[2336666]: [2336666] LOG:  invalid record length at 2/9F5CEE68: wanted 24, got 0
Nov 10 15:43:06 azazel postgres[2336666]: [2336666] LOG:  redo done at 2/9F5CEE28 system usage: CPU: user: 0.00 s, system: 0.00 s, elapsed: 0.00 s
Nov 10 15:43:06 azazel postgres[2336019]: [2336019] LOG:  database system is ready to accept connections
```

Reverting this whole PR works as a workaround.

Reply:
I think if the system call filter is the issue, you should see an error in the kernel log (or auditd, I'm not entirely sure).

But given that this is v8, there might be another problem: can you try setting `MemoryDenyWriteExecute` to false?

Reply:
> then we could just bisect through the list of changes?

I guess I would've just turned off more and more flags until it works. But yeah.

@jtojnar never used this extension, but any chance you can share a reproducer?

Reply:
Even just calling `ALTER EXTENSION plv8 UPDATE;` was enough to trigger the crash for me. Maybe the example select from https://plv8.github.io/ would work too. Not sure if I will be able to create a test today, maybe sometimes during the week.

Reply:
dmesg | grep audit

Reply:
Can you run `dmesg | grep audit` so we can see which syscall is required?

Reply:
A quick check gives me

```
vm # [   30.985080] audit: type=1326 audit(1731254105.832:2): auid=4294967295 uid=71 gid=71 ses=4294967295 subj=kernel pid=939 comm="postgres" exe="/nix/store/wc1a06ip2fajrjkfbw7cvxzw1c949a6g-postgresql-16.4/bin/postgres" sig=31 arch=c000003e syscall=330 compat=0 ip=0x7f0d72d04dfb code=0x80000000
```

Syscall 330 is `pkey_alloc` on x86_64-linux, so I think we'll need `@pkey` in the system call filter declaration. Haven't checked if the MemoryDenyWriteExecute stuff is also a a thing here, I think I've seen that with nodejs processes before, so we may want to check.

Can't promise that I'll be able to file a PR tonight, perhaps you'll have to wait until tomorrow.

Reply:
https://github.com/NixOS/nixpkgs/pull/355010

Reply:
Okay, tried adding `SystemCallFilter = [ "@pkey" ];` and got the errno fatal error above. Adding `MemoryDenyWriteExecute = false;` as well made it work.

Reply:
Do you know what triggers the second error? Would be nice to be able to repro.

Reply:
It is a pretty big SQL query including custom window functions:

<details>
<summary>Original query</summary>

```sql
WITH
  recent_event AS (
    SELECT
      id
    FROM
      event
    WHERE
      level = 'world'
      AND "end" < CURRENT_DATE
      AND "end" > '2022-01-01'
    ORDER BY
      "end" DESC
    LIMIT
      2
  ),
  past_team AS (
    SELECT
      *,
      0.5 AS coef,
      6 AS CEILING,
      position_in_class (age, 'ultraveteran') OVER (
        w
        ORDER BY
          score DESC,
          time ASC
      ) AS position_uv,
      count(nullif(is_in_class (age, 'ultraveteran'), FALSE)) OVER w AS limit_uv,
      position_in_class (age, 'superveteran') OVER (
        w
        ORDER BY
          score DESC,
          time ASC
      ) AS position_sv,
      count(nullif(is_in_class (age, 'superveteran'), FALSE)) OVER w AS limit_sv,
      position_in_class (age, 'veteran') OVER (
        w
        ORDER BY
          score DESC,
          time ASC
      ) AS position_v,
      count(nullif(is_in_class (age, 'veteran'), FALSE)) OVER w AS limit_v,
      position_in_class (age, 'open') OVER (
        w
        ORDER BY
          score DESC,
          time ASC
      ) AS position_o,
      count(nullif(is_in_class (age, 'open'), FALSE)) OVER w AS limit_o,
      position_in_class (age, 'junior') OVER (
        w
        ORDER BY
          score DESC,
          time ASC
      ) AS position_j,
      count(nullif(is_in_class (age, 'junior'), FALSE)) OVER w AS limit_j
    FROM
      team
    WHERE
      event_id IN (
        SELECT
          *
        FROM
          recent_event
      )
    WINDOW
      w AS (
        PARTITION BY
          event_id,
          gender
      )
  ),
  past_team_crit AS (
    SELECT
      *,
      (
        position_uv <= CEILING
        AND position_uv <= ceil(limit_uv * coef)
      ) AS prequalified_uv,
      (
        position_sv <= CEILING
        AND position_sv <= ceil(limit_sv * coef)
      ) AS prequalified_sv,
      (
        position_v <= CEILING
        AND position_v <= ceil(limit_v * coef)
      ) AS prequalified_v,
      (
        position_o <= CEILING
        AND position_o <= ceil(limit_o * coef)
      ) AS prequalified_o,
      (
        position_j <= CEILING
        AND position_j <= ceil(limit_j * coef)
      ) AS prequalified_j
    FROM
      past_team
  )
SELECT
  *
FROM
  member
  LEFT JOIN past_team_crit ON member.event_id = past_team_crit.event_id
  AND member.team_id = past_team_crit.id
WHERE
  prequalified_uv
  OR prequalified_sv
  OR prequalified_v
  OR prequalified_o
  OR prequalified_j;
```

</details>

I managed to pare it down slightly but it is not clear what is causing it:

<details>
<summary>Reduced query</summary>

```sql
WITH
  recent_event AS (
    SELECT
      id
    FROM
      event
    WHERE
      level = 'world'
      AND "end" < CURRENT_DATE
      AND "end" > '2022-01-01'
    ORDER BY
      "end" DESC
    LIMIT
      2
  ),
  past_team AS (
    SELECT
      *,
      0.5 AS coef,
      6 AS CEILING,
      position_in_class (age, 'ultraveteran') OVER (
        w
        ORDER BY
          score DESC,
          time ASC
      ) AS position_uv,
      count(nullif(is_in_class (age, 'ultraveteran'), FALSE)) OVER w AS limit_uv
    FROM
      team
    WHERE
      event_id IN (
        SELECT
          *
        FROM
          recent_event
      )
    WINDOW
      w AS (
        PARTITION BY
          event_id,
          gender
      )
  )
SELECT
  *,
  (
    position_uv <= CEILING
    AND position_uv <= ceil(limit_uv * coef)
  ) AS prequalified_uv
FROM
  past_team
```

</details>

Curiously, the crash triggers if I remove ultraveteran category or if I only keep ultraveteran category, but not when I keep any other category. I should really move the logic out of SQL.

Reply:
Wow, that is probably too big for a reproducer in a nixos test. Either way https://github.com/NixOS/nixpkgs/pull/355010 should be ready to go. Feel free to pick that into your tree, if that fits your workflow.

Reply:
Okay, I managed to make PLV8 (presumably) enable JIT with the following:

```sql
DO $$
let xs = [];
for (let i = 0, n = 4000000000; i < n; i++) {
    xs.push(Math.round(Math.random() * n))
}console.log(Math.sum(xs));
$$ LANGUAGE plv8;
```